src/xml2_url.cpp

```diff
@@ -1,156 +1,133 @@
+    out[i] = Xml2String(xmlBuildURI(uri, base_uri)).asRString();
   }
 
-  UNPROTECT(1);
   return out;
 }
 
-// [[export]]
-extern "C" SEXP url_relative_(SEXP x_sxp, SEXP base_sxp) {
-  R_xlen_t n = Rf_xlength(x_sxp);
-  SEXP out = PROTECT(Rf_allocVector(STRSXP, n));
-
-  if (Rf_xlength(base_sxp) > 1) {
-    Rf_error("Base URL must be length 1");
-  }
+[[cpp11::register]]
+cpp11::strings url_relative_(cpp11::strings x_sxp, cpp11::strings base_sxp) {
+  int n = x_sxp.size();
+  cpp11::writable::strings out(n);
 
-  const xmlChar* base_uri = (xmlChar*) Rf_translateCharUTF8(STRING_ELT(base_sxp, 0));
+  const xmlChar* base_uri = to_xml_chr(base_sxp, "Base URL");
 
   for (int i = 0; i < n; ++i) {
-    const xmlChar* uri = (xmlChar*) Rf_translateCharUTF8(STRING_ELT(x_sxp, i));
-    SET_STRING_ELT(out, i, Xml2String(xmlBuildRelativeURI(uri, base_uri)).asRString());
+    const xmlChar* uri = (xmlChar*) Rf_translateCharUTF8(x_sxp[i]);
+    out[i] = Xml2String(xmlBuildRelativeURI(uri, base_uri)).asRString();
   }
 
-  UNPROTECT(1);
   return out;
 }
 
-// [[export]]
-extern "C" SEXP url_parse_(SEXP x_sxp) {
-  R_xlen_t n = Rf_xlength(x_sxp);
+[[cpp11::register]]
+cpp11::data_frame url_parse_(cpp11::strings x_sxp) {
+  int n = x_sxp.size();
 
-  SEXP scheme = PROTECT(Rf_allocVector(STRSXP, n));
-  SEXP server = PROTECT(Rf_allocVector(STRSXP, n));
-  SEXP user = PROTECT(Rf_allocVector(STRSXP, n));
-  SEXP path = PROTECT(Rf_allocVector(STRSXP, n));
-  SEXP query = PROTECT(Rf_allocVector(STRSXP, n));
-  SEXP fragment = PROTECT(Rf_allocVector(STRSXP, n));
+  cpp11::writable::strings scheme(n);
+  cpp11::writable::strings server(n);
+  cpp11::writable::strings user(n);
+  cpp11::writable::strings path(n);
+  cpp11::writable::strings query(n);
+  cpp11::writable::strings fragment(n);
 
-  SEXP port = PROTECT(Rf_allocVector(INTSXP, n));
+  cpp11::writable::integers port(n);
 
   for (int i = 0; i < n; ++i) {
-    const char* raw = Rf_translateCharUTF8(STRING_ELT(x_sxp, i));
+    const char* raw = Rf_translateCharUTF8(x_sxp[i]);
     xmlURI* uri = xmlParseURI(raw);
     if (uri == NULL) {
       continue;
     }
 
-    SET_STRING_ELT(scheme, i, Rf_mkChar(uri->scheme == NULL ? "" : uri->scheme));
-    SET_STRING_ELT(server, i, Rf_mkChar(uri->server == NULL ? "" : uri->server));
-    INTEGER(port)[i] = uri->port == 0 ? NA_INTEGER : uri->port;
-    SET_STRING_ELT(user, i, Rf_mkChar(uri->user == NULL ? "" : uri->user));
-    SET_STRING_ELT(path, i, Rf_mkChar(uri->path  == NULL ? "" : uri->path));
-    SET_STRING_ELT(fragment, i, Rf_mkChar(uri->fragment == NULL ? "" : uri->fragment));
+    scheme[i] = uri->scheme == NULL ? "" : uri->scheme;
+    server[i] = uri->server == NULL ? "" : uri->server;
+    port[i] = uri->port == 0 ? NA_INTEGER : uri->port;
+    user[i] = uri->user == NULL ? "" : uri->user;
+    path[i] = uri->path  == NULL ? "" : uri->path;
+    fragment[i] = uri->fragment == NULL ? "" : uri->fragment;
```

Comment:
This one might get significantly slower, but again it doesn't _seem_ like it is used all that often? I couldn't really say

Reply:
I also think it shouldn't be used that often.